### PR TITLE
add foamMonitor to sanity checks of OpenFOAM

### PR DESCRIPTION
**Pull request by lexming**
Enhancement of the sanity checks of OpenFOAM related to PR https://github.com/easybuilders/easybuild-easyconfigs/pull/11770
Depends on ~~https://github.com/easybuilders/easybuild-easyconfigs/pull/11800~~ and ~~https://github.com/easybuilders/easybuild-easyconfigs/pull/11801~~

### CODE REVIEW

**Comment by lexming** [comment]
* Test with `OpenFOAM-8-foss-2020a.eb` from `develop` not having `gnuplot` as dependency:
https://gist.github.com/lexming/9f35f78eb3e466b88508cd29459a71b3
The build fails in the sanity check step as intended.

* Test with `OpenFOAM-8-foss-2020a.eb` from https://github.com/easybuilders/easybuild-easyconfigs/pull/11770 with `gnuplot` as dependency:
https://gist.github.com/lexming/377c6c45a98cb8ca5991f31dea272945
The build succeeds as intended.

**Comment by boegel** [comment]
@lexming There's a couple more things to check: how about older OpenFOAM versions, or variants (like OpenFOAM-Extend)?

Also, we can't merge this before adding the missing `gnuplot` to all OpenFOAM easyconfigs...

File: easybuild/easyblocks/o/openfoam.py [hunk 1]
```diff
@@ -447,7 +447,17 @@ def sanity_check_step(self):
             'dirs': dirs,
         }
 
-        super(EB_OpenFOAM, self).sanity_check_step(custom_paths=custom_paths)
+        # test the OpenFOAM environment in BASH
+        load_openfoam_env = "source $FOAM_BASH"
+        # test monitorFoam (exit code will only be 0 if all dependencies are met)
+        test_monitorFoam = " && ".join([load_openfoam_env, "foamMonitor -help"])
```
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py:453]: @lexming `OpenFOAM-Extend` doesn't provide `foamMonitor`, so this needs to be done conditionally

Done in https://github.com/lexming/easybuild-easyblocks/pull/3

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by boegel** [comment]
One additional complication: with OpenFOAM v2006, `foamMonitor -help` indeed complains when `gnuplot` is not found, but even if it is there, it always exits with exit code `1`:

```
== 2020-11-29 12:58:18,658 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/base/exceptions.py:124 in __init__): Sanity check failed: sanity check command source $FOAM_BASH && foamMonitor -help exited with code 1 (output:
Usage: foamMonitor [OPTION] <file>
options:
  -h | -help            prints the usage
  -i | -idle <time>     stops if <file> unchanging for <time> sec (default = 60)
  -l | -logscale        plots data (y-axis) on log scale, e.g. for residuals
  -r | -refresh <time>  refreshes display every <time> sec (default = 10)
  -y | -yrange <range>  sets data (y-axis) <range>, format "[0:1]"
  -g | -grid            draws grid lines on the plot

Monitor data with Gnuplot from time-value(s) graphs written by OpenFOAM
e.g. by functionObjects
- requires gnuplot, gnuplot_x11

Example:
  foamMonitor -l postProcessing/residuals/0/residuals.dat
```

So we should run `foamMonitor -help`, check whether it included "`Usage: foamMonitor`" (to check it includes the help output), but also that it *doesn't* include something like "`Gnuplot not installed`" or "`which: no gnuplot in`", at least for the versions where `foamMonitor -help` always exits with non-zero exit code...

**Comment by boegel** [comment]
More complications: the oldest OpenFOAM versions don't include `foamMonitor`:

```
sanity check command source $FOAM_BASH && foamMonitor -help exited with code 127
(output: /bin/bash: foamMonitor: command not found)
```

**Comment by lexming** [comment]
That complicates things indeed. OpenFOAM v2.2 and older do not have `foamMonitor`. I'll add a version check as well.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenFOAM-8-foss-2020a.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node374.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/5765d5e80caffb507f3581843e1e2785 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenFOAM-7-foss-2019b.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node357.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/ccca51d9d39b4ff796d4679648e2b209 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _OpenFOAM-2.4.0-intel-2019a.eb_ (partial log available at https://gist.github.com/9bb16ff8e6a4a57ee536acb0d75d4f4b)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
node357.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/ea52d3f0f7d91ebacf8a1bca40dd302d for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenFOAM-2.2.x-intel-2019a.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node357.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/6a4f04276cff158bb7e20d1e296c5d8c for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenFOAM-2.3.1-intel-2019b.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node357.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/3564a713ad3d30ee0e4abb8fbb4a2b27 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _OpenFOAM-6-foss-2019b.eb_ (partial log available at https://gist.github.com/8f558e63c7a6498bb30e9ad0be379deb)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
node373.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/09234e4850748a056b5124e415f50ca0 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _OpenFOAM-v1606+-foss-2018b.eb_ (partial log available at https://gist.github.com/df206de8921ef31786c759b7e009f1cf)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
node364.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/c67e058eafca3a4a62a6ca63cd7b7b07 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _OpenFOAM-v2006-foss-2020a.eb_ (partial log available at https://gist.github.com/80ddf8963f9d310e2add731935eda3a5)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
node364.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/b08592084b580bf1aa33af0063a99c9d for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _Qt-4.8.7-foss-2017b.eb_ 
 * **SUCCESS** _CGAL-4.11-foss-2017b-Python-2.7.14.eb_ 
 * **SUCCESS** _ParaView-5.4.1-foss-2017b-mpi.eb_ 
 * **FAIL (build issue)** _OpenFOAM-5.0-foss-2017b.eb_ (partial log available at https://gist.github.com/95ecd311a649b5bf06c45a96814e15f5)

Build succeeded for 3 out of 4 (1 easyconfigs in total)
node369.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/8749e641c5d961dd1b8fd37b1e3b527f for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _Mesquite-2.3.0-GCCcore-8.3.0.eb_ 
 * **SUCCESS** _ParMGridGen-1.0-gompi-2019b.eb_ 
 * **FAIL (build issue)** _OpenFOAM-Extend-4.1-20200408-foss-2019b-Python-2.7.16.eb_ (partial log available at https://gist.github.com/8f52e00c57210aa2d8b92dd30d69ed06)

Build succeeded for 2 out of 3 (1 easyconfigs in total)
node377.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/0ca54ad0c1501fad96671ee8e5a1b69a for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _OpenFOAM-v1906-foss-2019b.eb_ (partial log available at https://gist.github.com/626406a85e421dd1b7ae09ddafbfd755)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
node381.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/376b518471236b43f8f7441eb0a22e95 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenFOAM-Extend-4.1-20200408-foss-2019b-Python-2.7.16.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node373.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/cd65251af4219be601a050782b52e53e for a full test report.

**Comment by lexming** [comment]
This PR now works as intended. All easyconfigs of OpenFOAM fail or pass the sanity checks as expected. Currently, the only ones passing are:
* OpenFOAM <= v2.3, which do not have `foamMonitor`
* OpenFOAM v7 and v8, which have been already updated with `gnuplot`
* OpenFOAM-Extend, as it does not have `foamMonitor`

Next step, add the dependency on `gnuplot` to all easyconfigs of OpenFOAM that need it.

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenFOAM-v1812-foss-2018b.eb_ 
 * **SUCCESS** _OpenFOAM-v1906-foss-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-v1912-foss-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-v1912-intel-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-v2006-foss-2020a.eb_ 
 * **SUCCESS** _OpenFOAM-v2012-foss-2020a.eb_ 
 * **SUCCESS** _OpenFOAM-2.3.1-intel-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-2.4.0-intel-2019a.eb_ 
 * **SUCCESS** _OpenFOAM-Extend-4.0-intel-2019b-Python-2.7.16.eb_ 
 * **SUCCESS** _OpenFOAM-Extend-4.1-20191120-intel-2019b-Python-2.7.16.eb_ 
 * **SUCCESS** _OpenFOAM-Extend-4.1-20200408-foss-2019b-Python-2.7.16.eb_ 
 * **SUCCESS** _OpenFOAM-6-foss-2018b.eb_ 
 * **SUCCESS** _OpenFOAM-6-foss-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-6-intel-2018a.eb_ 
 * **SUCCESS** _OpenFOAM-7-foss-2019b-20200508.eb_ 
 * **SUCCESS** _OpenFOAM-7-foss-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-8-foss-2020a.eb_ 

Build succeeded for 17 out of 17 (17 easyconfigs in total)
node3104.skitty.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/93a3b730671b1053859322bc614d44b9 for a full test report.

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenFOAM-Extend-4.1-20200408-foss-2019b-Python-2.7.16.eb_ 
 * **SUCCESS** _OpenFOAM-2.2.x-intel-2019a.eb_ 
 * **SUCCESS** _OpenFOAM-2.4.0-intel-2019a.eb_ 
 * **SUCCESS** _OpenFOAM-3.0.1-intel-2018a.eb_ 
 * **SUCCESS** _OpenFOAM-5.0-20180108-foss-2018a.eb_ 
 * **SUCCESS** _OpenFOAM-5.0-20180108-intel-2017b.eb_ 
 * **SUCCESS** _OpenFOAM-5.0-20180108-intel-2018a.eb_ 
 * **SUCCESS** _OpenFOAM-5.0-foss-2017b.eb_ 
 * **SUCCESS** _OpenFOAM-5.0-intel-2017b.eb_ 
 * **SUCCESS** _OpenFOAM-6-foss-2018b.eb_ 
 * **SUCCESS** _OpenFOAM-6-foss-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-6-intel-2018a.eb_ 
 * **SUCCESS** _OpenFOAM-7-foss-2019b-20200508.eb_ 
 * **SUCCESS** _OpenFOAM-7-foss-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-8-foss-2020a.eb_ 
 * **SUCCESS** _OpenFOAM-v1606+-foss-2018b.eb_ 
 * **SUCCESS** _OpenFOAM-v1612+-foss-2018b.eb_ 
 * **SUCCESS** _OpenFOAM-v1712-foss-2017b.eb_ 
 * **SUCCESS** _OpenFOAM-v1712-intel-2017b.eb_ 
 * **SUCCESS** _OpenFOAM-v1806-foss-2018b.eb_ 
 * **SUCCESS** _OpenFOAM-v1812-foss-2018b.eb_ 
 * **SUCCESS** _OpenFOAM-v1906-foss-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-v1912-foss-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-v1912-intel-2019b.eb_ 
 * **SUCCESS** _OpenFOAM-v2006-foss-2020a.eb_ 
 * **SUCCESS** _OpenFOAM-v2012-foss-2020a.eb_ 

Build succeeded for 26 out of 26 (26 easyconfigs in total)
node2666.swalot.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (haswell), Python 3.6.8
See https://gist.github.com/4c27ac68f885d4129c800210554ad2ca for a full test report.

**Pull request merged**